# Choosing PING model parameters for a _Peak Patch_ simulation 
This script is an aid for selecting appropriate PING model parameters to simulate a region of the universe of specified size with specified resolution using the _Peak Patch_ simulations.

In [7]:
import numpy as np
import astropy
import astropy.units as u
import camb
from astropy.cosmology import Planck18,z_at_value
import matplotlib.pyplot as plt
plt.rcParams['text.usetex'] = True

The PING model parameters are as follows

\begin{align*}
	\epsilon &= \frac{M_P^2}{2} \left( \frac{V'}{V} \right)^2
	\\
	&= \frac{M_P^2}{2} \left[ \frac{1}{V} \frac{d}{d\phi} \left( 3M_P^2H^2 - \frac{\dot{\phi}^2}{2} \right) \right]^2
\end{align*}

In a general, multi-field inflation model, we could consider the inflaton $\phi$ and one dominant uncorrelated field $\chi$. $\chi$ is independent of $\phi$ and thus transverse in the phase space of the potential $V(\phi,\chi)$. For an instability in an inflaton domain $\phi \in (\phi_p-\phi_w, \phi_p+\phi_w)$ the inflationary potential is the regular scalar field potential plus a term due to the instability
\begin{equation} \label{eq: inflationary potential}
	V(\phi,\chi) = \frac{1}{2} m_\phi^2\phi^2 + \frac{1}{2} m_\chi^2\chi^2 + \Delta V(\phi,\chi)
\end{equation}
where $m_\phi$ and $m_\chi$ are the respective masses of the inflationary scalar fields (heavier fields having stronger $k$ dependence at low $k$) and the change due to the instability is
\begin{equation} \label{eq: change in inflationary potential due to instability}
	\Delta V(\phi,\chi) = \left\{
	\begin{matrix}
		\frac{\lambda_\chi}{4} \left[ \left( \frac{\phi-\phi_p}{\phi_w} \right)^2 -1 \right]^2 \left[ \left( \chi^2-v^2 \right)^2 - v^4 \right]
		&& \forall \phi \in (\phi_p-\phi_w, \phi_p+\phi_w)
		\\
		0 && \forall \phi \not\in (\phi_p-\phi_w, \phi_p+\phi_w)
	\end{matrix}
	\right.
\end{equation}
For fixed $\phi \in (\phi_p-\phi_w, \phi_p+\phi_w)$, $\Delta V$ is quartic in $\chi$ with a local maximum $\Delta V(\phi,\chi=0)=0$ and a pair of global minima at points $\chi=\pm v$, where $v$ is referred to as the vacuum expectation value of $\chi$ as it represents the lowest energy state that $\chi$ can have. Taking $\phi=\phi_p$ the perturbation becomes $\Delta V(\phi_p,\chi) = \frac{\lambda_\chi}{4}\chi^4 - \frac{1}{2} m_\lambda^2 \chi^2$, and we see from the first term that $\lambda_\chi$ is a quartic coupling of the potential to $\chi$. In the second term we recognize that the coefficient for a quadratic coupling is like a mass so we define the tachyonic mass<a name="ref_1"></a>[<sup>[1]</sup>](#footnote_1) $m_\lambda = \sqrt{\lambda_\chi} v$.

The depth of the minima is greatest at $\Delta V(\phi_p,\pm v) = -\frac{\lambda_\chi  v^4}{4}$ and goes to zero at $\phi=\phi_p\pm\phi_w$, respectively the start ($+$) and end ($-$) of the instability. The latter occurs at a primordial redshift $z_e = a_e^{-1}-1 \gg 0$. Because the inflation model is not fully described, exactly when the instability occurs is not known. This means both that we don't know how far to extend the potential beyond $\phi\in(\phi_p-\phi_w, \phi_p+\phi_w)$ and at what scale factor the instability occurs. Because the lattice simulaitons are not a full simulation of inflation, $\phi_p$ and $a_e$ are degenerate free parameters, so we can fix $\phi_p$ and vary $a_e$.

For every point on the CMB sky to have been in causal contact prior to inflation, this would require a $\gtrsim 60$ $e$-fold expansion during inflation, with most models predicting at least 70 $e$-foldings (meaning the the scale factor changes by a factor of $a_f/a_i \sim e^{70} \simeq 2.5\cdot10^{30}$). %The instability is a means of bringing about the end of inflation, so it will presumably come toward the end of the inflaitonary period, but there is a lot of play in $a_e$. 

We use an integrator to solve for the power spectrum of the transverse inflationary field at the end of the instability $\chi_e$ over some range of wave numbers $k_e\in[\Delta k_e,k_{e,\text{max}}]$. A cubic simulation with spatial comoving volume $s_\text{box}^3$ discretized into $n^3$ cubic voxels has a discrete set of wave numbers<a name="ref_2"></a>[<sup>[2]</sup>](#footnote_2) running from $\Delta k_e = 2\pi s_\text{box}^{-1}$ to $k_{e,\text{max}} = \sqrt{3}n\pi s_\text{box}^{-1}$.

%Wave numbers scale as $s_\text{box}^{-1}$, so the primordial fundamental wavenumber must be $\Delta k_e \le 2\pi / a_e s_\text{box}$, and the maximum wavenumber in such a volume at primordial time must be $k_{e,\text{max}} \ge (n+2)\pi/a_e s_\text{box}$.

The early-universe simulations use naturalized units with $\hbar=c=1$ (see appendix \ref{sec: appC}) so lengths are expressed in units of one over mass $\bar\ell = \hbar c^{-1} m^{-1} = m^{-1}$, and comoving<a name="ref_3"></a>[<sup>[3]</sup>](#footnote_3) wave numbers $\hbar^{-1}c \bar{k}$, whith units of mass. Masses are expressed in terms of a value $\mu = 10^{-5} M_{Pl} = 10^{-5} (8\pi)^{-1/2} m_P$ where the Planck mass is $m_P$ as in (\ref{eq: Planck Units}), so the wavenumber that appears in the code is $\bar{k}_\text{code} = \mu^{-1} \bar{k}$. The scale factor $a'$ in the lattice simulations is set so that $a'_s=1$ at the start of the instability, whereas \emph{Peak Patch} uses a scale factor $a$ with $a_0=1$ at present. The comoving wavenumber in \emph{Peak Patch} conventions at the end of the instability $k_e$ is therefore% from the lattice simulations $\bar{k}_\text{code}$ can be found
\begin{align*}
	k_e &= a_e (a'_e)^{-1} \hbar^{-1}c \bar{k}
	\\
	k_e &= a_e (a'_e)^{-1} \hbar^{-1}c \bar{k}_\text{code} \mu
	\\
	k_e &= a_e (a'_e)^{-1} 10^{-5} (8\pi)^{-1/2} \hbar^{-1} c m_P \bar{k}_\text{code}
	\\
	k_e &= a_e (a'_e)^{-1} 10^{-5} (8\pi)^{-1/2} \ell_P^{-1} \bar{k}_\text{code}
\end{align*}
Here we recognize the Planck length $\ell_P$, also defined in (\ref{eq: Planck Units}). Whereas \emph{Peak Patch} uses units of $\{\text{Mpc},M_\odot,\text{s}\}$, so it is convenient to use a Planck length in what is perhaps a rather silly-looking choice of units $\ell_P = \hbar^{1/2} G^{1/2} c^{-3/2} = 5.2379258(59)\cdot10^{-58} \text{ Mpc}$ (where the number in parentheses is the one-standard-deviation uncertainty in the last two digits as per NIST 2018 CODATA). 
\begin{equation}
	k_e
	%= a_e e^{-\alpha'_e} \ell_\text{code}^{-1} \bar{k}_\text{code}
	= a_e (a'_e)^{-1} \ell_\text{code}^{-1} \bar{k}_\text{code},
	\qquad
	\ell_\text{code} 
	\equiv 10^5 (8\pi)^{1/2} \ell_P
	= 2.625907(30)\cdot10^{-52} \text{ Mpc}
\end{equation}
The lattice-simulation scale factor can be read from the code and has a value $a_e' \sim \mathcal{O}(0)$. \emph{Peak Patch} uses wave numbers on the order of hundreds of kiloparsecs to tens of gigaparsecs, so $a_e$ must be extremely small $\mathcal{O}(-50)$ or so if the features are to be on LSS scales.

As an example, an $(10\text{ Gpc})^3$ box in \emph{Peak Patch} with a resolution of $n=10^4$ would require a comoving wavenumber of about $k_0 \in [2\pi\cdot10^{-4}, 2\pi]$. So an intermediate wavenumber $k_c = 2\pi\cdot10^{-2}$. The primordial power $P_{\chi\chi}(k)$ has a dropoff at about $\bar{k}_\text{code} \sim 10^2$ and $a'_e \sim 3$ so that gives $a_e \sim 4.9\cdot10^{-55}$.





So for a _Peak Patch_ run, the minumum and maximum wavenumbers that will have $k_\mathrm{min} = 2\pi s_\mathrm{box}^{-1}$ and $k_\mathrm{max} = \sqrt{3} n \pi s_\mathrm{box}^{-1}$.

In [8]:
s_box = 8200.0 # Gpc
n_eff = 8200

k_min = 2*np.pi/s_box
k_max = 3**.5*n_eff*np.pi/s_box

print(k_min)
print(k_max)

0.0007662421106316568
5.441398092702653


The main run deometries I've been using are:
| run       | $s_\mathrm{box}$    | $n$    | $k_\mathrm{min}$     | $k_\mathrm{max}$ | $a_{e,\mathrm{min}}$  | $a_{e,\mathrm{max}}$  |
| ---       | ---                 | ---    | ---                  | ---              | ---                   | --- |
| WebSky2.0 | $8200~\mathrm{Gpc}$ | $8200$ | $7.662\times10^{-4}$ | $5.441$          | $6.036\times10^{-57}$ | $4.286\times10^{-53}$ |
| LIM debug | $500~\mathrm{Gpc}$  | $1000$ | $1.257\times10^{-2}$ | $10.88$          | $9.902\times10^{-56}$ | $8.571\times10^{-53}$

I've typically been using $m_\phi=m_\chi=1$, $\phi_w=0.12547$, $m_\lambda=25.0$ and $v=0.1$ which yields a primordial power $P_{\chi\chi}(k)$ with a dropoff at about $\bar{k}_\mathrm{code} \sim 10^2$ and $a_e'\sim 3$.

I've mostly been using $a_e = 10^{-53}$ which is close to the maximum allowed value for both boxes, so we want to experiment with other values. This means running _Peak Patch_ with those and seeing what the power spectra look like. We don't want an overall large shift, so probably $m_\lambda$ will need to be shifted concurrent with $a_e$, which in turn means we need to double check that the cutoff scale $k_\mathrm{cut}$ remains within the desired $k$ range.

In [10]:
a_e_prime = 3
l_code    = 2.625907e-52 # Mpc
k_code    = 1e2

k_e       = 10.88
a_e       = k_e * a_e_prime * l_code * k_code**-1

print(a_e)

8.570960448e-53


<a name="ref_1"></a>[1] [^](#footnote_1) If $m_\lambda^2 > m_\chi^2$ then the total mass in the term quadratic in $\chi$ of the full potential $\frac{1}{2}(m_\chi^2-m_\lambda^2)\chi^2$ becomes imaginery, which is the tachyon solution, hence we refer to $m_\lambda = \sqrt{\lambda_\chi}v$ as the tachyonic mass.

<a name="ref_2"></a>[2] [^](#footnote_2) The available wave numbers are given by each unique distance between lattice cells in the discretized Fourier-space field
\begin{equation*}
	k_{j_1,j_2,j_3} = \frac{2 \pi \sqrt{j_1^2 + j_2^2 + j_3^2}}{s_\text{box}},
	\qquad
	%j_1 \le j_2 \le j_3,
	j_1,j_2,j_3 \in [1,n/2+1) \cap \mathbb{Z}
\end{equation*}
The wave number folds at the Nyquist frequency (half the sample frequency) along each axis, so the largest value of $j$ is $n/2$, which gives the maximum value $\sqrt{3}n\pi s_\text{box}^{-1}$.

Note that power spectra are better suited to log-log plots, so we tend to use log-spaced wave number bins in discrete $P(k)$ tables and then interpolate to get a given $k_{j_1,j_2,j_3}$. In that case, $N$ bins ranging from $k_1=\Delta k_e$ to $k_N=k_{e,\text{max}}$ are $k_j=k_1(k_N/k_1)^{j/N}$.

<a name="ref_3"></a>[3] [^](#footnote_3) The FRW metric is $ds^2 = dt^2 - a^2(t) \delta_{ij} dq^i dq^j$. The spatial coordinates $\mathbf{q}$ are called comoving because they expand with the expanding spacetime. The distance you would actually measure are in terms of Eulerian coordinates $\mathbf{x} = a \mathbf{q}$. A comoving wave number is constructed from comoving distance scales $k = 2\pi/|\Delta\mathbf{q}|$. The physically relevant wave number is the Eulerian wave number though, constructed from Eulerian distance scales $k/a = 2\pi/|\Delta\mathbf{x}|$.